### DISCUSSION TITLE
nodejs standard modules and polyfills?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#5395] [Q&A]
I'm interested in starting a repo that attempts to duplicate the nodejs environment as closely as possible by providing standard module implementations and polyfills. Have you given this much thought?

One thing I noticed right away is that `require('foo')` won't resolve to `pb_hooks/foo.js`. It needs to be 

```
require(`${__hooks}/foo`)
```

What about making require's root dir `__hooks`? Is there any reason not to? Or perhaps a `--jsModulesDir` config?

Edit: I started https://github.com/benallfree/pocketbase-node

**Comment by benallfree** [comment]
Example...

```js
const readFileSync = (path, options) => {
  const res = $os.readFile(path);
  if (typeof res === "string") {
    return res;
  }
  const s = String.fromCharCode.apply(null, res);
  return s;
};

module.exports = { readFileSync };
```

This kind of thing along with something to polyfill Buffer etc would be great.

Also, a `console.log( { foo } )` that doesn't just output `Object [object]` :D

If any of this is interesting, I'm happy to send some core PR's as well. We could do a lot of this as JS extensions, but a few Go mods might be helpful unless you'd like to bake standard nodejs modules into PB.

**Reply by TokenChingy** [reply]
Would it be better to add the implementations to [goja_nodejs](https://github.com/dop251/goja_nodejs) and then update [https://github.com/pocketbase/pocketbase/blob/master/plugins/jsvm/jsvm.go](https://github.com/pocketbase/pocketbase/blob/master/plugins/jsvm/jsvm.go)?

**Reply by benallfree** [reply]
It is missing a lot of core modules, unfortunately :(

But yes in theory another way would be to contribute to that project instead.
